### ISSUE TITLE
pipeline basic steps build/publish docker image

### ISSUE THREAD

**Issue by PXzzzpython** [jenkinsci/docker-plugin#737]
If you are sure that this plugin is the source of your troubles then please report:
 - [ ] docker-plugin version you use: Pipeline: Basic Steps v2.18
 - [ ] jenkins version you use: jenkins v2.176.1
 - [ ] docker engine version you use: 18.06.1-ce
 - [ ] details of the docker container(s) involved and details of how the docker-plugin is connecting to them

![image](https://user-images.githubusercontent.com/35281450/59582248-3d8f1b00-910a-11e9-83d9-9fcd15ee3fab.png)

this is my jenkinsfile
`stage('build image') {
        step([$class: 'DockerBuilderPublisher', cleanImages: false, cleanupWithJenkinsJobDelete: false, cloud: 'docker', dockerFileDirectory: '', fromRegistry: [], pushCredentialsId: 'xxx', pushOnSuccess: true, tagsString: 'xxx'])
    }`

- [ ] stack trace / logs / any technical details that could help diagnose this issue

`java.lang.ClassCastException: com.nirima.jenkins.plugins.docker.builder.DockerBuilderPublisher.fromRegistry expects class org.jenkinsci.plugins.docker.commons.credentials.DockerRegistryEndpoint but received class java.util.ArrayList
	at org.jenkinsci.plugins.structs.describable.DescribableModel.coerce(DescribableModel.java:446)
	at org.jenkinsci.plugins.structs.describable.DescribableModel.buildArguments(DescribableModel.java:366)
	at org.jenkinsci.plugins.structs.describable.DescribableModel.instantiate(DescribableModel.java:288)
Caused: java.lang.IllegalArgumentException: Could not instantiate {cloud=docker, cleanupWithJenkinsJobDelete=false, dockerFileDirectory=, fromRegistry=[], pushOnSuccess=true, cleanImages=false, pushCredentialsId=xxxxxx, tagsString=xxx for DockerBuilderPublisher(dockerFileDirectory: String, fromRegistry: DockerRegistryEndpoint(url: String, credentialsId: String), cloud: String, tagsString: String, pushOnSuccess: boolean, pushCredentialsId: String, cleanImages: boolean, cleanupWithJenkinsJobDelete: boolean)
	at org.jenkinsci.plugins.structs.describable.DescribableModel.instantiate(DescribableModel.java:293)
	at org.jenkinsci.plugins.structs.describable.DescribableModel.coerce(DescribableModel.java:428)
	at org.jenkinsci.plugins.structs.describable.DescribableModel.buildArguments(DescribableModel.java:366)
	at org.jenkinsci.plugins.structs.describable.DescribableModel.instantiate(DescribableModel.java:288)
	at org.jenkinsci.plugins.workflow.steps.StepDescriptor.newInstance(StepDescriptor.java:205)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeStep(DSL.java:262)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeMethod(DSL.java:176)
	at org.jenkinsci.plugins.workflow.cps.CpsScript.invokeMethod(CpsScript.java:122)
	at sun.reflect.GeneratedMethodAccessor298.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1213)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:42)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:158)
	at org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:23)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:157)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:156)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:160)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:130)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:130)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:17)
Caused: java.lang.IllegalArgumentException: Could not instantiate {delegate={$class=DockerBuilderPublisher, cleanImages=false, cleanupWithJenkinsJobDelete=false, cloud=docker, dockerFileDirectory=, fromRegistry=[], pushCredentialsId=xxxxxx, pushOnSuccess=true, tagsString=xxxxxx for CoreStep(delegate: SimpleBuildStep{ArtifactArchiver(artifacts: String, allowEmptyArchive?: boolean, caseSensitive?: boolean, defaultExcludes?: boolean, excludes?: String, fingerprint?: boolean, onlyIfSuccessful?: boolean) | DockerBuilderControl(option: DockerBuilderControlOption{DockerBuilderControlOptionProvisionAndStart(cloudName: String, templateId: String) | DockerBuilderControlOptionRun(cloudName: String, image: String, pullCredentialsId: String, dnsString: String, network: String, dockerCommand: String, volumesString: String, volumesFrom: String, environmentsString: String, hostname: String, memoryLimit: int, memorySwap: int, cpuShares: int, shmSize: int, bindPorts: String, bindAllPorts: boolean, privileged: boolean, tty: boolean, macAddress: String) | DockerBuilderControlOptionStart(cloudName: String, containerId: String) | DockerBuilderControlOptionStop(cloudName: String, containerId: String, remove: boolean) | DockerBuilderControlOptionStopAll(remove: boolean)}) | DockerBuilderPublisher(dockerFileDirectory: String, fromRegistry: DockerRegistryEndpoint(url: String, credentialsId: String), cloud: String, tagsString: String, pushOnSuccess: boolean, pushCredentialsId: String, cleanImages: boolean, cleanupWithJenkinsJobDelete: boolean) | Fingerprinter(targets: String) | GitHubCommitNotifier(resultOnFailure: String, statusMessage?: ExpandableMessage(content: String)) | GitHubCommitStatusSetter(commitShaSource?: GitHubCommitShaSource{BuildDataRevisionShaSource() | ManuallyEnteredShaSource(sha: String)}, contextSource?: GitHubStatusContextSource{DefaultCommitContextSource() | ManuallyEnteredCommitContextSource(context: String)}, errorHandlers?: StatusErrorHandler{ChangingBuildStatusErrorHandler(result: String) | ShallowAnyErrorHandler()}[], reposSource?: GitHubReposSource{AnyDefinedRepositorySource() | ManuallyEnteredRepositorySource(url: String)}, statusBackrefSource?: GitHubStatusBackrefSource{BuildRefBackrefSource() | ManuallyEnteredBackrefSource(backref: String)}, statusResultSource?: GitHubStatusResultSource{ConditionalStatusResultSource(results: ConditionalResult{AnyBuildResult(message?: String, state?: String) | BetterThanOrEqualBuildResult(message?: String, result?: String, state?: String)}[]) | DefaultStatusResultSource()}) | GitHubSetCommitStatusBuilder(contextSource?: GitHubStatusContextSource{DefaultCommitContextSource() | ManuallyEnteredCommitContextSource(context: String)}, statusMessage?: ExpandableMessage(content: String)) | JUnitResultArchiver(testResults: String, allowEmptyResults?: boolean, healthScaleFactor?: double, keepLongStdio?: boolean, testDataPublishers?: TestDataPublisher{}[]) | JavadocArchiver(javadocDir: String, keepAll: boolean) | Mailer(recipients: String, notifyEveryUnstableBuild: boolean, sendToIndividuals: boolean) | WsCleanup(cleanWhenAborted?: boolean, cleanWhenFailure?: boolean, cleanWhenNotBuilt?: boolean, cleanWhenSuccess?: boolean, cleanWhenUnstable?: boolean, cleanupMatrixParent?: boolean, deleteDirs?: boolean, disableDeferredWipeout?: boolean, externalDelete?: String, notFailBuild?: boolean, patterns?: Pattern(pattern: String, type: PatternType[INCLUDE, EXCLUDE])[], skipWhenFailed?(deprecated): boolean)})
	at org.jenkinsci.plugins.structs.describable.DescribableModel.instantiate(DescribableModel.java:293)
	at org.jenkinsci.plugins.workflow.steps.StepDescriptor.newInstance(StepDescriptor.java:205)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeStep(DSL.java:262)
	at org.jenkinsci.plugins.workflow.cps.DSL.invokeMethod(DSL.java:176)
	at org.jenkinsci.plugins.workflow.cps.CpsScript.invokeMethod(CpsScript.java:122)
	at sun.reflect.GeneratedMethodAccessor298.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1213)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at org.codehaus.groovy.runtime.callsite.PogoMetaClassSite.call(PogoMetaClassSite.java:42)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:158)
	at org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:23)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:157)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:156)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:160)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:130)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:130)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:17)
	at WorkflowScript.run(WorkflowScript:23)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:57)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:109)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:82)
	at sun.reflect.GeneratedMethodAccessor296.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.CollectionLiteralBlock$ContinuationImpl.dispatch(CollectionLiteralBlock.java:55)
	at com.cloudbees.groovy.cps.impl.CollectionLiteralBlock$ContinuationImpl.item(CollectionLiteralBlock.java:45)
	at sun.reflect.GeneratedMethodAccessor297.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
	at com.cloudbees.groovy.cps.Next.step(Next.java:83)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:174)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:163)
	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:129)
	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:268)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:163)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:18)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:51)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:186)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:370)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$200(CpsThreadGroup.java:93)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:282)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:270)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:64)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:131)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:59)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)`

**Comment by pjdarton** [comment]
The Jenkinsfile is specifying `fromRegistry: []`, so the Jenkinsfile is supplying an empty list for the fromRegistry argument.
The exception message says `fromRegistry expects class org.jenkinsci.plugins.docker.commons.credentials.DockerRegistryEndpoint but received class java.util.ArrayList`
i.e. it's complaining that you're giving it an `ArrayList` when it needs a `DockerRegistryEndpoint`, and you're giving it a (empty) list.

 - TL;DR: the error message is correct; the Jenkinsfile is incorrect.

Note: If you want to use pipelines & docker, you're better off using the [docker-pipeline-plugin]( https://wiki.jenkins.io/display/JENKINS/Docker+Pipeline+Plugin) as documented [here](https://jenkins.io/doc/book/pipeline/docker/).  The docker-plugin (i.e. this plugin) wasn't designed with pipelines in mind.

**Comment by PXzzzpython** [comment]
> The Jenkinsfile is specifying `fromRegistry: []`, so the Jenkinsfile is supplying an empty list for the fromRegistry argument.
> The exception message says `fromRegistry expects class org.jenkinsci.plugins.docker.commons.credentials.DockerRegistryEndpoint but received class java.util.ArrayList`
> i.e. it's complaining that you're giving it an `ArrayList` when it needs a `DockerRegistryEndpoint`, and you're giving it a (empty) list.
> 
> * TL;DR: the error message is correct; the Jenkinsfile is incorrect.
> 
> Note: If you want to use pipelines & docker, you're better off using the [docker-pipeline-plugin](https://wiki.jenkins.io/display/JENKINS/Docker+Pipeline+Plugin) as documented [here](https://jenkins.io/doc/book/pipeline/docker/). The docker-plugin (i.e. this plugin) wasn't designed with pipelines in mind.

Thank you very much. I already know how to solve the problem

**Comment by pjdarton** [comment]
OK, I'm confused now - I thought you were asking for assistance.
If you're happy that this isn't a bug in the code then we can close the issue (and it'll help anyone who hits the same problem, searches, and finds this issue text).

If I've misunderstood and you believe there is a bug, please clarify what you think is wrong (and what it should be instead) & feel free to re-open the issue.
